### PR TITLE
[UHN-BS-20] Server now pushing notifications

### PR DESCRIPTION
**Pull request by Asiapenguin**
- Reorganized code
  - utils was basically services: moved Online Service
- Extracted User Service
  - Changed code to use service wherever it reduced lines of code
  - Note: The service does not use lean() (i.e. can still call save() on the same variable)
- Created Notification Service
- Help Request schema
- User schema
  - Removed version key
  - Added push_token attribute
- Help Request controller
  - POST /help_request to create a help request
    - Sends notification via Notification Service
- User controller
  - Added POST /user/{id}/notifications
    - sends device token to backend for storage (stored in push_token attribute)
- Documentation
- Testing endpoints and controller
  - You can send a notification via an endpoint, provided the user you use has a device token.

### CODE REVIEW

**Comment by Asiapenguin** [comment]
This is also connected to #39

File: Controllers/user.js [hunk 1]
```diff
@@ -237,22 +216,24 @@ async function addResponders(req, res) {
...
+  var user = null;
+
+  try {
+    user = await UserService.findUserById(req.params.id);
+  } catch (err) {
+    handle.notFound(res, err.message);
   }
+
+  var responders = user.get("responders");
+  let hasResponderID = responders.some(
```
Comment (ad2969) [inline Controllers/user.js]:
```suggestion
  const hasResponderID = responders.some(
```

File: README.md [hunk 2]
```diff
@@ -138,3 +138,52 @@ HTTP Response:
...
+    "status": enum["open", "sent_to_responder", "taken", "arrived", "resolved"] (initialized to "open"),
+    "responders: [
+        { id: string }
+        ...
+    ],
+    "created_at": date,
+    "updated_at": date
+}
+```
+## Adding Help Request
```
Comment (ad2969) [inline README.md]:
```suggestion
## Adding Push Token
```

File: services/notification.service.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+var expo = expo_server.getExpoInstance();
+
+// TODO: Need to handle receipts of notifications, to handle errors
+const sendBatchNotifications = async (responders, help_request) => {
+  let push_tokens = [];
+  let notifications = [];
+  let user = null;
+
+  try {
```
Comment (ad2969) [inline services/notification.service.js]: include userdata and stuff in sendBatchNotif as an argument

File: README.md [hunk 4]
```diff
@@ -138,3 +138,52 @@ HTTP Response:
...
+
+## Adding Help Request
+
+POST "/help_request"
+
+Request Body:
+
+```
+{
+    "user_id": string
```
Comment (mxnyee) [inline README.md]: Frontend has never used snake_case up to this point. A lot of the other endpoints take request data in camelCase too

File: server.js [hunk 5]
```diff
@@ -32,6 +36,12 @@ app.get("/users/:id/responders", middleware.checkToken, user.getResponders);
 app.get("/users/:id/responders/count",middleware.checkToken,user.getResponderCount);
 app.delete("/users/:id/responders/:responderid", middleware.checkToken, user.deleteResponder);
 
+// Help requests
+app.post("/help_requests", middleware.checkToken, help_request.addHelpRequest);
```
Comment (mxnyee) [inline server.js]: Is it weird to have underscores in URLs?
```suggestion
app.post("/help-requests", middleware.checkToken, help_request.addHelpRequest);
```

File: README.md (hunk 4 above)
Comment (Asiapenguin) [reply README.md]: Changed to camel. Check other attributes as well

File: Controllers/user.js [hunk 6]
```diff
@@ -91,38 +73,35 @@ async function signupUser(req, res) {
...
+  try {
+    user = await UserService.findUserById(req.params.id);
+  } catch (err) {
+    handle.notFound(res, err.message);
   }
+
+  var onlineStatus = await OnlineService.checkOnlineStatus(user._id);
+
+  let result = UserService.cleanUser(user);
+  result.online_status = onlineStatus;
```
Comment (mxnyee) [inline Controllers/user.js]:
```suggestion
  result.onlineStatus = onlineStatus;
```

File: README.md [hunk 7]
```diff
@@ -138,3 +138,52 @@ HTTP Response:
     "count": int
 }
 ```
+
+## Adding Help Request
+
+POST "/help_request"
```
Comment (mxnyee) [inline README.md]:
```suggestion
POST "/help-request"
```

File: Controllers/user.js [hunk 8]
```diff
@@ -1,11 +1,12 @@
 let jwt = require("jsonwebtoken");
 var bcrypt = require("bcrypt");
 var ObjectId = require("mongodb").ObjectId;
-var handle = require("../Utils/error_handling");
-const { customValidationResult } = require("../Utils/error_handling");
+var handle = require("../utils/error_handling");
```
Comment (ad2969) [inline Controllers/user.js]: I've sorta learned to think that "const" should be what's normally used when calling "require" or doing imports and stuff. I would suggest it but its likely that Im just being petty tho

File: Models/user.js [hunk 9]
```diff
@@ -14,8 +14,9 @@ const model = mongoose.model("users", new Schema({
     },
     note: String,
   },
-  responders: [{_id:false, id: String}]
-}));
+  responders: [{_id:false, id: String}],
+  pushToken: Object
```
Comment (matthewleemle) [inline Models/user.js]: Why is pushToken an Object is the request is a string?

File: Controllers/user.js (hunk 8 above)
Comment (Asiapenguin) [reply Controllers/user.js]: It is a good point to use const instead for these import statements

File: controllers/help_request.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    responders: responders
+  });
+
+  try {
+    await help_request.save();
+  } catch (err) {
+    handle.internalServerError(res, "Cannot create help request.");
+  }
+
+  await NotificationService.sendBatchNotifications(user, help_request);
```
Comment (ad2969) [inline controllers/help_request.js]: This is all still a "TODO" right

Comment (Asiapenguin) [reply controllers/help_request.js]: If you are asking about the sendBatchNotifications, yes. It still needs to handle potential errors coming back as a Promise in the ticket chunks. We need that data for analytics as well later (# notifications that were sent successfully / # responders contacted)

File: Controllers/user.js [hunk 11]
```diff
@@ -15,51 +16,35 @@ async function loginUser(req, res) {
...
-            }
-          );
+    try {
+      if (bcrypt.compareSync(password, result.password)) {
+        let token = jwt.sign({ id: result._id }, process.env.SECRET, {
+          expiresIn: "24h"
+        });
 
-          OnlineService.setOnline(result._id.toString());
+        console.log("token: ", token);
```
Comment (ad2969) [inline Controllers/user.js]: Looks like token is alrdy logged  in the res.json() call~
```suggestion
```

File: Models/user.js (hunk 9 above)
Comment (Asiapenguin) [reply Models/user.js]: Good question. I initially thought the token would be an object but turns out to be something like ExponentPushToken[blahblahblah]. I will turn it back into a string type

**Comment by Asiapenguin** [comment]
@ad2969 @mxnyee Is this good to go?

File: Controllers/user.js [hunk 12]
```diff
@@ -211,7 +189,7 @@ async function addResponders(req, res) {
 
         for (var i = 0, len = respondersToAdd.length; i < len; i++) {
           user.responders.push(respondersToAdd[i]);
-          user.save();
+          await user.save();
```
Comment (mxnyee) [inline Controllers/user.js]: as a note, I feel like this _might_ solve (or have something to do with) https://github.com/CodetheChangeFoundation/UHN-Backend-Server/issues/50

File: server.js [hunk 13]
```diff
@@ -20,6 +22,8 @@ app.post("/login", validateLogin(), user.loginUser);
 
 app.post("/users/:id/responders", middleware.checkToken, user.addResponders);
 
+app.post("/users/:id/notificationToken", user.addPushToken);
```
Comment (mxnyee) [inline server.js]: this is picky but can we do
```suggestion
app.post("/users/:id/notification-token", user.addPushToken);
```
to be consistent

File: Controllers/user.js (hunk 12 above)
Comment (mxnyee) [reply Controllers/user.js]: @matthewleemle

File: server.js [hunk 14]
```diff
@@ -32,6 +36,12 @@ app.get("/users/:id/responders", middleware.checkToken, user.getResponders);
 app.get("/users/:id/responders/count",middleware.checkToken,user.getResponderCount);
 app.delete("/users/:id/responders", middleware.checkToken, user.deleteResponders);
 
+// Help requests
+app.post("/help-requests", middleware.checkToken, help_request.addHelpRequest);
```
Comment (mxnyee) [inline server.js]: Hold on shouldn't it be
```suggestion
app.post("/help-request", middleware.checkToken, help_request.addHelpRequest);
```
?

Comment (Asiapenguin) [reply server.js]: Nope, this is fine, its appending one help request to a collection of help requests.

Comment (mxnyee) [reply server.js]: readme needs to be changed then. also in all your pull requests you wrote “/help-request”, which is why i pointed it out

Comment (Asiapenguin) [reply server.js]: Oh must have been a typo, my bad!

**Comment by Asiapenguin** [comment]
@mxnyee Approve if you think its ready and I will merge

**Pull request merged**